Add Cloud Task Queue resource

mccall114 opened the pull request:
New resource `google_task_queue` as part of issue https://github.com/terraform-providers/terraform-provider-google/issues/2367.

First time contributor here. I read the guidelines about splitting out new packages in a separate PR, and I'm happy to do that if this one is ok.

mccall114 commented:
Linter failed in travis. Running locally with go 1.11.13 confirms that it's failing on code I didn't touch.

bamItsCam commented on google/resource_task_queue.go:
```diff
@@ -0,0 +1,241 @@
+}
+
+func resourceTaskQueueImportState(d *schema.ResourceData, meta interface{}) ([]*schema.ResourceData, error) {
+	parts := strings.Split(d.Id(), "/")
```
Consider doing what resource_kms_key_ring.go does for its import, using parseImportId() and replaceVars().

tysen commented on google/resource_task_queue.go:
```diff
@@ -0,0 +1,342 @@
+				MaxItems: 1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"max_burst_size": {
```
The API docs say this field is output only so we can omit it from the schema.

tysen commented on google/resource_task_queue.go:
```diff
@@ -0,0 +1,342 @@
+					},
+				},
+			},
+			"retry": {
```
I would prefer if this matched the API: `retry_config`

tysen commented on google/resource_task_queue.go:
```diff
@@ -0,0 +1,342 @@
+				MaxItems: 1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"host": {
```
Also output only.

tysen commented on google/resource_task_queue_test.go:
```diff
@@ -0,0 +1,276 @@
+						"google_task_queue.fizzbuzz", location, queueName, &queue),
+					testAccCheckTaskQueueEquals(
+						"google_task_queue.fizzbuzz", &queue),
+				),
```
I don't think we need these checks (nor the funcs, and I think that also means we can avoid vendoring this new `context` dependency) - the `ImportStateVerify` is sufficient.

mccall114 commented:
@tysen, I've removed those tests and renamed `retry` to `retry_config`. I left `max_burst_size` and `host` so that we could still see them in the output. LMK what you think.

tysen commented on google/resource_task_queue_test.go:
```diff
@@ -0,0 +1,149 @@
+
+		expectedName := fmt.Sprintf("projects/%s/locations/%s/queues/%s", config.Project, attributes["location"], attributes["name"])
+
+		_, err := config.clientCloudTasks.Projects.Locations.Queues.Get(expectedName).Context(context.Background()).Do()
```
Why include this call to `Context`? Does this not work without it like the other services?

mccall114 replied:
Sorry for the delay, I was on vacation. I've removed the call to `Context`.

jba commented:
I'd love to start using this. What is holding it up? Can I help?

tysen commented:
Still waiting for a response to the question in my review.

pokojskis commented:
I would really love to use it

tysen commented:
PR looks good now. If you could submit separate PRs (one for TPG and one for TPGB) to vendor the new dependency (and then rebase this PR to remove the vendoring here), afterwards I can upstream this PR into MM and get it merged.

mccall114 commented:
@tysen, I've rebased this branch and opened https://github.com/terraform-providers/terraform-provider-google/pull/4781
and https://github.com/terraform-providers/terraform-provider-google-beta/pull/1315.

tysen commented:
@mccall114 Looks like you need to sign the CLA https://github.com/GoogleCloudPlatform/magic-modules/pull/2587

mccall114 commented:
@tysen I just signed it

tysen commented:
@mccall114 Thanks, will you post `@googlebot I fixed it.` on that PR?

seriousManual commented:
any updates on this?

tysen commented:
I had some issues finalizing this implementation, and it'll be easier to maintain and extend if it's generated by Magic Modules, so I submitted https://github.com/GoogleCloudPlatform/magic-modules/pull/2662 which is under review.

tysen commented:
Should be in the next release: https://github.com/GoogleCloudPlatform/magic-modules/pull/2662

mccall114 commented:
Thanks @tysen!
